### PR TITLE
Add TypeScript support. Closes #1025

### LINKED ISSUE #1025: Add built-in TypeScript support
The existing `--types` support is designed for JS projects with an exported types definition file. This will add support for using lab directly with TS projects without the need to use the [lab-transform-typescript](https://www.npmjs.com/package/lab-transform-typescript) module.

### PR DESCRIPTION
**Pull request by hueniverse**

### CODE REVIEW

**Review by devinivy — APPROVED** [review]
Was finally able to give this a proper read.  It looks good to me and I'm all for the functionality— thanks!

**Review by kanongil — COMMENTED** [review]
It would be nice with better typescript integration, but I'm concerned about officially supporting it. The main issue is that the transpiling can introduce extra checks in the output, some of which can be impossible to reach using tests, causing coverage to fall. I don't know if there is a reasonable solution to this?

I would also like to see a test for how `--typescript` behaves when the dependency is missing, similar to https://github.com/hapijs/lab/blob/65333961eb348c04da3150b38e1f4d1e8c1b5205/test/runner.js#L1967

**Comment by hueniverse** [comment]
> It would be nice with better typescript integration, but I'm concerned about officially supporting it. The main issue is that the transpiling can introduce extra checks in the output, some of which can be impossible to reach using tests, causing coverage to fall. I don't know if there is a reasonable solution to this?

We need to move to use v8 built-in coverage tooling to get proper coverage reports and get rid of the way we do it today. That said, it works pretty well. I haven't hit any actual issues with coverage so far. The additional code is pretty small and supporting it would be trivial. If limitations are found in the future, we can deal with them then or move support elsewhere as last resort.

File: lib/modules/leaks.js [hunk 1]
```diff
@@ -128,15 +128,45 @@ const internals = {
     symbols: [
         Symbol.toStringTag,
         Symbol.for('@hapi/lab/coverage/_state')
+    ],
+
+    typescript: [
```
[the remark this answers is not in the document]
Comment (hueniverse) [reply lib/modules/leaks.js:133]: We already maintain this for node. When new ones show up, we add them. Not a big deal.

File: lib/modules/typescript.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+const Typescript = require('typescript');
+
+const internals = {
+    configs: {}
+};
+
+internals.transform = function (content, fileName) {
+
+    const path = Typescript.normalizePath(Typescript.sys.getCurrentDirectory());
+    const configFile = process.env.TSCONFIG || Typescript.findConfigFile(path, Typescript.sys.fileExists);
```
[the remark this answers is not in the document]
Comment (hueniverse) [reply lib/modules/typescript.js:12]: Fixed using proposed replacement.

File: lib/modules/typescript.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    if (!internals.configs[configFile]) {
```
[the remark this answers is not in the document]
Comment (hueniverse) [reply lib/modules/typescript.js:14]: Done.

File: lib/modules/typescript.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (hueniverse) [reply lib/modules/typescript.js:12]: Done.

File: lib/modules/typescript.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        const { config, error } = Typescript.readConfigFile(configFile, Typescript.sys.readFile);
+        if (error) {
+            console.error(`TypeScript config error in ${configFile}: ${error.messageText}`);
+            process.exit(1);
```
[the remark this answers is not in the document]
Comment (hueniverse) [reply lib/modules/typescript.js:18]: I copied this from the cli but I agree that since this isn't in the cli we should throw instead.

**Comment by hueniverse** [comment]
> I would also like to see a test for how --typescript behaves when the dependency is missing, similar to

No need since this is only in the cli, not the reporter so it will just fail as expected.

**Comment by kanongil** [comment]
> > It would be nice with better typescript integration, but I'm concerned about officially supporting it. The main issue is that the transpiling can introduce extra checks in the output, some of which can be impossible to reach using tests, causing coverage to fall. I don't know if there is a reasonable solution to this?
> 
> We need to move to use v8 built-in coverage tooling to get proper coverage reports and get rid of the way we do it today. That said, it works pretty well. I haven't hit any actual issues with coverage so far. The additional code is pretty small and supporting it would be trivial. If limitations are found in the future, we can deal with them then or move support elsewhere as last resort.

I'm talking the coverage tool when applied to `.ts` files through the transform. It will only see the resulting `.js` code, which can contain more statements than in the `.ts` code.  Eg. the generated code for a declared `enum` has an untestable conditional code path, as discussed in #1002.

**Pull request merged**